File: examples/skills/cek_stok.js
#!/usr/bin/env node
/**
 * Skill: cek_stok (Node.js version)
 * Checks product stock levels by calling your system's REST API.
 *
 * Test locally:
 *   echo '{"produk":"Indomie"}' | \
 *     CLIENT_API_BASE_URL=http://localhost:3000 \
 *     CLIENT_API_AUTH="Bearer mytoken" \
 *     node examples/skills/cek_stok.js
 */

const http = require('http')
const https = require('https')
const { URL, URLSearchParams } = require('url')

async function main() {
  const params = JSON.parse(require('fs').readFileSync('/dev/stdin', 'utf8'))
  const produk = (params.produk || '').trim()
  const gudang = (params.gudang || '').trim()

  if (!produk) {
    respond([], "Parameter 'produk' is required.")
    return
  }

  const baseUrl = process.env.CLIENT_API_BASE_URL || ''
  const auth    = process.env.CLIENT_API_AUTH || ''

  const url = new URL(`${baseUrl}/products/stock`)
  url.searchParams.set('search', produk)
  if (gudang) url.searchParams.set('warehouse', gudang)

  try {
    const body = await get(url.toString(), auth)
    const items = body.data || (Array.isArray(body) ? body : [])
    respond(items, buildSummary(produk, items))
  } catch (err) {
    respond([], `Error: ${err.message}`)
  }
}

function get(url, auth) {
  return new Promise((resolve, reject) => {
    const client = url.startsWith('https') ? https : http
    const options = { headers: {} }
    if (auth) options.headers['Authorization'] = auth

    client.get(url, options, res => {
      let data = ''
      res.on('data', chunk => { data += chunk })
      res.on('end', () => {
        if (res.statusCode >= 400) {
          reject(new Error(`HTTP ${res.statusCode}: ${data}`))
          return
        }
        try {
          resolve(JSON.parse(data))
        } catch {
          reject(new Error('Invalid JSON response from API'))
        }
      })
    }).on('error', reject)
  })
}

function buildSummary(produk, items) {
  if (!items.length) return `No products found matching '${produk}'.`
  const lines = [`Found ${items.length} product(s):`]
  for (const item of items) {
    const name      = item.name || item.nama || 'Unknown'
    const sku       = item.sku || '-'
    const qty       = item.qty ?? item.stock ?? 0
    const warehouse = item.warehouse || item.gudang || ''
    let line = `- ${name} (SKU: ${sku}): ${qty} units`
    if (warehouse) line += ` at ${warehouse}`
    lines.push(line)
  }
  return lines.join('\n')
}

function respond(data, summary) {
  process.stdout.write(JSON.stringify({ data, summary }))
}

main().catch(err => {
  process.stdout.write(JSON.stringify({ data: [], summary: `Unexpected error: ${err.message}` }))
})

File: examples/skills/cek_stok.py
#!/usr/bin/env python3
"""
Skill: cek_stok (REST API version)
Checks product stock levels by calling your system's REST API.

Customize:
  - The API endpoint path (/products/stock)
  - The response field names (name, sku, qty, warehouse)

Test locally:
  echo '{"produk": "Indomie"}' | \
    CLIENT_API_BASE_URL=http://localhost:3000 \
    CLIENT_API_AUTH="Bearer mytoken" \
    python3 examples/skills/cek_stok.py
"""

import json
import os
import sys
import urllib.error
import urllib.parse
import urllib.request


def main():
    params = json.load(sys.stdin)
    produk = params.get("produk", "").strip()
    gudang = params.get("gudang", "").strip()

    if not produk:
        respond([], "Parameter 'produk' is required.")
        return

    base_url = os.environ.get("CLIENT_API_BASE_URL", "")
    auth = os.environ.get("CLIENT_API_AUTH", "")

    # Build query — adjust field names to match your API
    query = urllib.parse.urlencode({k: v for k, v in {
        "search": produk,
        "warehouse": gudang,
    }.items() if v})

    url = f"{base_url}/products/stock?{query}"
    req = urllib.request.Request(url)
    if auth:
        req.add_header("Authorization", auth)

    try:
        with urllib.request.urlopen(req, timeout=10) as resp:
            body = json.loads(resp.read())
    except urllib.error.HTTPError as e:
        respond([], f"API returned error {e.code}: {e.reason}")
        return
    except urllib.error.URLError as e:
        respond([], f"Could not connect to system API: {e.reason}")
        return

    # Adjust field names to match your API response shape
    items = body.get("data", body if isinstance(body, list) else [])
    respond(items, build_summary(produk, items))


def build_summary(produk, items):
    if not items:
        return f"No products found matching '{produk}'."
    lines = [f"Found {len(items)} product(s):"]
    for item in items:
        # Adjust field names to match your data shape
        name = item.get("name") or item.get("nama", "Unknown")
        sku = item.get("sku", "-")
        qty = item.get("qty") or item.get("stock", 0)
        warehouse = item.get("warehouse") or item.get("gudang", "")
        line = f"- {name} (SKU: {sku}): {qty} units"
        if warehouse:
            line += f" at {warehouse}"
        lines.append(line)
    return "\n".join(lines)


def respond(data, summary):
    print(json.dumps({"data": data, "summary": summary}))


if __name__ == "__main__":
    main()
